src/librbd/librbd.cc

```diff
@@ -3340,6 +3367,72 @@ extern "C" ssize_t rbd_list_children(rbd_image_t image, char *pools,
   return ret;
 }
 
+extern "C" int rbd_list_children2(rbd_image_t image,
+                                  rbd_child_info_t *children,
+	                          int *max_children)
+{
+  vector<librbd::child_info_t> cpp_children;
+  librbd::ImageCtx *ictx = (librbd::ImageCtx *)image;
+  tracepoint(librbd, list_children_enter, ictx, ictx->name.c_str(), ictx->snap_name.c_str(), ictx->read_only);
+
+  if (!max_children) {
+    tracepoint(librbd, list_children_exit, -EINVAL);
+    return -EINVAL;
+  }
+
+  int r = librbd::list_children(ictx, &cpp_children);
+  if (r == -ENOENT) {
+    tracepoint(librbd, list_children_exit, *max_children);
+    return 0;
```

Comment:
Nit: if ```r == -NOENT```, just set ```r = 0``` and continue so that ```max_return``` is properly populated and the array is properly null terminated.